### PR TITLE
Added a kagi search extension

### PR DESCRIPTION
**Pull request by fearoffish**
## Description

This adds a search (with suggestions) extension to Raycast, for [kagi.com](https://kagi.com). Invites to the search are available on their site. This extension requires a token you are provided when you sign up.

This is a difficult one to test, because you need to be in on the beta.

## Type of change

- New extension

## Screencast

https://cln.sh/AmvPUA

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

### CODE REVIEW

File: kagi-search/package.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      "description": "Kagi search with autosuggestions",
+      "mode": "view"
+    }
+  ],
+  "preferences": [
+    {
+      "name": "kagiToken",
+      "type": "textfield",
+      "title": "Kagi Token",
+      "description": "Please enter your Kagi Token",
```
Comment (es183923) [inline kagi-search/package.json:22]: `Token` should be `token`

File: kagi-search/package.json [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      "name": "index",
+      "title": "Kagi Search",
+      "description": "Kagi search with autosuggestions",
+      "mode": "view"
+    }
+  ],
+  "preferences": [
+    {
+      "name": "kagiToken",
+      "type": "textfield",
```
Comment (es183923) [inline kagi-search/package.json:20]: use `password` instead

**Comment by es183923** [comment]
There should be some information on how to get your token.

**Comment by fearoffish** [comment]
and the extension should be in the correct folder. Wow, what a terrible PR. My apologies, I'll fix it.

File: extensions/kagi-search/package.json [hunk 3]
```diff
@@ -17,9 +17,9 @@
   "preferences": [
     {
       "name": "token",
-      "type": "private",
+      "type": "password",
       "title": "Kagi Token",
-      "description": "Please enter your Kagi Token",
+      "description": "Please enter your Kagi token, which can be found at https://kagi.com/settings?p=user_details in your Session Link",
```
Comment (es183923) [inline extensions/kagi-search/package.json:22]: This can just be `Found at https://kagi.com/settings?p=user_details in your Session Link`

**Comment by es183923** [comment]
Is `list-icon.png` used? If not, it should be deleted.

**Comment by es183923** [comment]
Also, `command-icon.png` should be 512x512, not 76x76.

**Comment by fearoffish** [comment]
Tested locally with changes and all is good.

**Comment by RishabSwift** [comment]
This is great!

Just feedback: I think having `search Kagi for ...` subtitle is a bit redundant since the user already knows that they're searching using Kagi, and the title also already displays the suggestions

ScreenShot 2022-01-29 at 03 04 07@2x

**Comment by fearoffish** [comment]
I think the solution is I change the "Results" title to be "Search Kagi for..." and then just list results.

**Comment by fearoffish** [comment]
Like so:

https://cloud.fearof.fish/ZOMG7y

File: extensions/kagi-search/src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  const { isLoading, history, results, searchText, search, addHistory, deleteAllHistory, deleteHistoryItem } =
+    useSearch(token);
+
+  const listItems: SearchResult[] = searchText.length === 0 ? history : results;
+
+  if (!token) {
+    showToast(ToastStyle.Failure, "Please add your Kagi token to this extension's preferences");
+    return <List>{null}</List>;
+  }
+
```
Comment (pernielsentikaer) [inline extensions/kagi-search/src/index.tsx:31]: Redundant since token can't be empty

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @fearoffish 👋 

Awesome extension, I left some feedback for you to address 💪 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

File: extensions/kagi-search/package.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  "name": "kagi-search",
+  "title": "Kagi Search",
+  "description": "Searches the Closed Beta of Kagi Search, using the autosuggestions url.",
+  "icon": "command-icon.png",
+  "author": "fearoffish",
+  "license": "MIT",
+  "commands": [
+    {
+      "name": "index",
+      "title": "Kagi Search",
```
Comment (pernielsentikaer) [inline extensions/kagi-search/package.json:12]:
```suggestion
      "title": "Search Kagi",
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Could you change this, and then we're ready to go 💪 Sorry for extra waiting time 😄 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

File: extensions/kagi-search/package.json (hunk 5 above)
Comment (fearoffish) [reply extensions/kagi-search/package.json:12]: The name of the search engine is Kagi Search, it's what they show in the HTML `<title>` too. Do you still suggest I change it?

Comment (fearoffish) [reply extensions/kagi-search/package.json:12]: Also, the Google Search extension has the same naming convention (I used that as my template). But, if you'd prefer I change it then I'm happy to.

Comment (fearoffish) [reply extensions/kagi-search/package.json:12]: Done

Comment (pernielsentikaer) [reply extensions/kagi-search/package.json:12]: > Done

Ah yeah, my bad, I had something else in my head here! Kagi Search is the right thing!

Apologies! 🙏

Comment (fearoffish) [reply extensions/kagi-search/package.json:12]: Actually, I thought about it...and I think *you're* right! I want to "Search Kagi" but the title of the extension is "Kagi Search". haha, let's go with the change :)

Comment (fearoffish) [reply extensions/kagi-search/package.json:12]: my brain when I want to use the extension says "let's **S**earch **K**agi" not the other way round.

Comment (pernielsentikaer) [reply extensions/kagi-search/package.json:12]: I think I confused myself again, hahaha, oh well 😂😂😂

**Comment by pernielsentikaer** [comment]
Hi @fearoffish 👋 

It seems like I can't find the command title correction, could you please provide it 😄

**Comment by fearoffish** [comment]
My mistake, I only pushed to my own extension repo!

**Review by pernielsentikaer — APPROVED** [review]
Hi @fearoffish 🧇 

Much better 🔥

**Pull request merged**